Toric/Mathlib/RingTheory/HopfAlgebra/Convolution.lean

```diff
@@ -134,47 +175,136 @@ lemma counit_comp_antipode :
+    · rw [map_counit_comp_comul_right]
+      simp only [coassoc_simps, LinearMap.convOne_def, Algebra.linearMap_tensorProduct]
+  · trans (LinearMap.mul' R (C ⊗[R] C) ∘ₗ (δ ⊗ₘ δ)) ∘ₗ (.id ⊗ₘ antipode R) ∘ₗ δ
+    · simp only [LinearMap.convMul_def, coassoc_simps]
+    trans (δ ∘ₗ Algebra.linearMap R C) ∘ₗ ε
+    · rw [Bialgebra.mul'_comp_map_comul_comul]
+      simp [coassoc_simps]
+    · congr
+      ext
+      simp [TensorProduct.one_def]
+
+open MulOpposite
+
+lemma map_antipode_antipode_comp_comul : (antipode R ⊗ₘ antipode R) ∘ₗ δ =
+    (TensorProduct.comm R C C).toLinearMap ∘ₗ δ ∘ₗ antipode R := by
+  rw [← map_antipode_antipode_comp_comul']
+  simp [coassoc_simps]
+
+/-- The antipode as a coalgebra hom. -/
+def antipodeCoalgHom [IsCocomm R C] : C →ₗc[R] C where
+  __ := antipode R
+  map_smul' := _
+  counit_comp := counit_comp_antipode
+  map_comp_comul := by
+    dsimp
+    rw [map_antipode_antipode_comp_comul, ← LinearMap.comp_assoc, comm_comp_comul]
+
+lemma LinearMap.algHom_comp_convOne
+    {R A B C : Type*} [CommSemiring R] [AddCommMonoid A] [Semiring B] [Semiring C]
+    [Module R A] [Coalgebra R A] [Algebra R B] [Algebra R C] (f : B →ₐ[R] C) :
+    f.toLinearMap.comp (1 : A →ₗ[R] B) = 1 := by
+  ext
+  exact (f : B →ₐ[R] C).commutes _
+
+lemma LinearMap.convOne_comp_coalgHom
+    {R A B C : Type*} [CommSemiring R] [AddCommMonoid A] [AddCommMonoid B] [Semiring C]
+    [Module R A] [Coalgebra R A] [Module R B] [Coalgebra R B]
+    [Algebra R C] (f : A →ₗc[R] B) :
+    (1 : B →ₗ[R] C).comp f.toLinearMap = 1 := by
+  ext
+  exact congr(algebraMap R C ($(f.counit_comp) _))
+
+lemma BialgHom.comp_antipode (f : A →ₐc[R] C) :
+    f.toLinearMap.comp (antipode R) = (antipode R).comp f.toLinearMap := by
+  apply left_inv_eq_right_inv (a := f.toLinearMap)
+  · refine (LinearMap.algHom_comp_convMul_distrib (f : A →ₐ[R] C) (antipode R) .id).symm.trans ?_
+    rw [LinearMap.antipode_mul_id, LinearMap.algHom_comp_convOne]
+  · refine (LinearMap.convMul_comp_coalgHom_distrib .id (antipode R) f.toCoalgHom).symm.trans ?_
+    rw [LinearMap.id_mul_antipode, LinearMap.convOne_comp_coalgHom]
+
+end Semiring
+section CommSemiring
+
+variable [CommSemiring A] [CommSemiring C] [HopfAlgebra R A] [HopfAlgebra R C]
 
 /-- The antipode as a coalgebra hom. -/
-def antipodeBialgHom : C →ₐc[R] C where
-  __ := antipodeAlgHom R (A := C)
+def antipodeBialgHom [IsCocomm R A] : A →ₐc[R] A where
+  __ := antipodeAlgHom R A
   map_smul' := _
   counit_comp := counit_comp_antipode
   map_comp_comul := by
-    apply left_inv_eq_right_inv (a := comul)
-    · sorry
-    · sorry
+    dsimp
+    rw [map_antipode_antipode_comp_comul, ← LinearMap.comp_assoc, comm_comp_comul]
+
+instance [IsCocomm R A] : Inv (C →ₐc[R] A) where inv := antipodeBialgHom.comp
+
+instance [IsCocomm R C] : Inv (C →ₐc[R] A) where
+  inv f := (f.comp (antipodeBialgHom)).copy (antipode R ∘ f) congr($(f.comp_antipode.symm))
```

Comment:
Oh god. Do we need something like [`SecondCountableTopologyEither`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Constructions/BorelSpace/Basic.html#SecondCountableTopologyEither) but for `IsCocomm`?